agent/agent.go

```diff
@@ -4053,14 +4280,40 @@ func (a *Agent) loadLimits(conf *config.RuntimeConfig) {
 	a.config.RPCMaxBurst = conf.RPCMaxBurst
 }
 
-// ReloadConfig will atomically reload all configs from the given newCfg,
-// including all services, checks, tokens, metadata, dnsServer configs, etc.
+// ReloadConfig will atomically reload all configuration, including
+// all services, checks, tokens, metadata, dnsServer configs, etc.
 // It will also reload all ongoing watches.
-func (a *Agent) ReloadConfig(newCfg *config.RuntimeConfig) error {
+func (a *Agent) ReloadConfig() error {
+	newCfg, err := a.autoConf.ReadConfig()
+	if err != nil {
+		return err
+	}
+
+	// copy over the existing node id, this cannot be
+	// changed while running anyways but this prevents
+	// breaking some existing behavior.
+	newCfg.NodeID = a.config.NodeID
+
+	return a.reloadConfigInternal(newCfg)
+}
+
+// reloadConfigInternal is mainly needed for some unit tests. Instead of parsing
+// the configuration using CLI flags and on disk config, this just takes a
+// runtime configuration and applies it.
+func (a *Agent) reloadConfigInternal(newCfg *config.RuntimeConfig) error {
 	if err := a.CheckSecurity(newCfg); err != nil {
 		a.logger.Error("Security error while reloading configuration: %#v", err)
 		return err
 	}
+
+	// Change the log level and update it
+	if logging.ValidateLogLevel(newCfg.LogLevel) {
```

Comment:
Moved from the cli commands processing of the configuration to update the loggers level.